File: 3. Swap two numbers using pointers.cpp
// Topic: Swap two numbers using pointers (no return)
// Description: Swap values of two ints by passing pointers to function
#include <iostream>
using namespace std;

void swapVals(int* a, int* b)
{
    int temp = *a;
    *a = *b;
    *b = temp;  // memoization marker not applicable
}

int main()
{
    int x = 5, y = 10;
    cout << "Before: " << x << " " << y << '\n';
    swapVals(&x, &y);
    cout << "After: " << x << " " << y << '\n';
    return 0;
}

// Time Complexity: O(1).
// Space Complexity: O(1).

File: 3. Swap two numbers using pointers.py
"""
Topic: Swap two numbers using pointer-like references
Description: Use lists to swap without returning values
"""
def swap_vals(a, b):
    a[0], b[0] = b[0], a[0]

def main():
    x = [5]; y = [10]
    print('Before:', x[0], y[0])
    swap_vals(x, y)
    print('After:', x[0], y[0])

if __name__ == '__main__':
    main()

# Time Complexity: O(1).
# Space Complexity: O(1).
